### DISCUSSION TITLE
Does BruteForceSampler pull parameters randomly or exhausts one first?

### DISCUSSION THREAD

**Discussion by simpsus** [optuna/optuna#5403] [Q&A]
basically title.

To be more precise: is it more

```
for p1 in p1_choices:
 for p2 in p2_choices:
  return p1, p2
```

or
```

p1 = random.choice(p1_choices)
p2 = random.choice(p2_choices)
return p1,p2 # need to check for duplicates still
```

**Comment by Daraan** [comment]
The BruteForceSampler does also sample randomly. But not in a naive way. For each parameter it puts the finite options of the distribution and puts it into a tree - one layer represents one parameter. For each trial one path (that is currently not running) is sampled until all paths in the tree have been chosen. This allows randomness while avoiding duplicates.
